docs/cudf/source/developer_guide/benchmarking.md

```diff
@@ -0,0 +1,323 @@
+Objects that differ in these ways should be interchangeable for most benchmarks.
+The goal of writing benchmarks in this way is to then automatically benchmark objects with different properties.
+We support this use case with the `benchmark_with_object` decorator.
+
+The use of this decorator is best demonstrated by example:
+
+```python
+@benchmark_with_object(cls="dataframe", dtype="int", cols=6)
+def bench_foo(benchmark, dataframe):
+    benchmark(dataframe.foo)
+```
+
+In the example above `bench_foo` will be run for DataFrames containing six columns of integer data.
+The decorator allows automatically parametrizing the following object properties:
+
+- `cls`: Objects of a specific class, e.g. `DataFrame`.
+- `dtype`: Objects of a specific dtype.
+- `nulls`: Objects with and without null entries.
+- `cols`: Objects with a certain number of columns.
+- `rows`: Objects with a certain number of rows.
+
+In the example, since we did not specify the number of rows or nullability,
+it will be run once for each valid number of rows and for both nullable and non-nullable data.
+The valid set of all parameters (e.g. the numbers of rows) is stored in the `common/config.py` file.
+This decorator allows a developer to write a generic benchmark that works for many types of objects,
+then have that benchmark automatically run for all objects of interest.
+
+### Parametrizing tests
+
+The `benchmark_with_object` decorator covers most use cases and automatically guarantees a baseline of benchmark coverage.
+However, many benchmarks will require more customized objects.
+In some cases those will be the primary targets whose methods are called.
+For instance, a benchmark may require a `Series` with a specific data distribution.
+In others, those objects will be arguments passed to other functions.
+An example of this is `DataFrame.where`, which accepts many types of objects to filter with.
+
+In the first case, fixtures should follow certain rules.
+When writing fixtures, developers should make the data sizes dependent on the benchmarks configuration.
+The `benchmarks/common/config.py` file defines standard data sizes to be used in benchmarks.
+These data sizes can be tweaked for debugging purposes (see [](#testing-benchmarks) below).
+Fixture sizes should be relative to the `NUM_ROWS` and/or `NUM_COLS` variables defined in the config module.
+These rules ensure consistency between these fixtures and those provided by `benchmark_with_object`.
+
+In the second case, fixtures are really functioning as parameters, which we discuss in the next section.
+
+
+### Parametrization: custom fixtures, `pytest-cases` and `pytest.mark.parametrize`
+
+When it comes to parametrizing tests, we have a number of options at our disposal.
+One option is fixtures, while a second is using `pytest.mark.parametrize`.
+A third option is provided by the [`pytest_cases`](https://smarie.github.io/python-pytest-cases/) `pytest` plugin.
```

Comment:
Did we come to a conclusion about whether we want to advocate for the use of `cases` v/s straight parametrize/fixtures? It's not super clear from the [thread](https://github.com/rapidsai/cudf/pull/11180). cc: @wence-

Reply:
My fear is that _unless_ it's cut-and-dried, most developers would rather not deal with complexity.

If I develop a feature and need to write a test or benchmark for it, I'm going to choose the path of least resistance, which could include:

- writing as little code as possible
- touching as few files as possible
- learning as few new tools as possible

Reply:
Sorry to pile on this discussion further (and late!).

I think the current set of rules for choosing a method of parametrization is well-written and as concise as it can be. 

But, it's quite mind-bending to think about the intersectionality of parameters for someone wanting to develop a feature and write incremental tests or benchmarks for it. Incorporating fixtures into this workflow is easy and familiar as one identifies reusable components between tests. Incorporating cases seems more difficult and really only possible in retrospect - or am I thinking of this all wrong?

Reply:
I am not really sufficiently familiar with cases to say, but I think that is reasonable. Probably you would start out writing fixtures, and then, as they became more baroque, wonder if there were a better way (which I believe cases offer).

Reply:
I don't think that anything that you've said is wrong or untrue, but I think you're classifying as a difference in kind something that is really only a difference in degree.

I would argue that incorporating fixtures effectively is also really only possible in retrospect. As our test suite clearly demonstrates, most people's starting point (at least historically) is to write an unparametrized test with some hardcoded objects, then abuse `pytest.mark.parametrize` to test as many possibilities as you can. Getting everyone to switch to using fixtures already requires some amount of either 1) foresight into what tests you want to write, or 2) refactoring tests in hindsight. Typically the latter occurs as part of PR review, although over the past year or two I think many developers have become more accustomed to thinking in terms of reusable fixtures and reaching for them first.

Cases are just one more step up the complexity ladder. Instead of just thinking about how you might write a bunch of tests that take the exact same arguments, in order to start off with using cases you need to start thinking about tests that might have only partial overlap in parametrization. Otherwise, you refactor tests in hindsight. The review process will train people in those best practices in the same way that we've improved our fixture usage.

> I think the current set of rules for choosing a method of parametrization is well-written and as concise as it can be.

Just to make sure that I understand: am I interpreting this as saying that you're not yet convinced that we should use cases at all?

Reply:
I did a poor job communicating what changes are required here.

I'm not unilaterally opposed to the use of `cases`, but I feel I need a better understanding of how much friction, if any, we would be adding to the process of developing benchmarks and - more importantly - tests, if we require them.

I think before moving forward with this recommendation, it'd be great to have a deeper discussion about `cases` offline. Perhaps with the rest of the team so we can get their feedback as well?

Reply:
FWIW, bringing my experience from other projects that heavily used parameterized fixtures (but not pytest cases). I think there's a tradeoff between generality (and lack of code repetition) that the more complex approaches bring, and ease of debugging (and extraction) of a particular test when something goes wrong and you need to fix things. Although it is relatively easy to run a single test with `pytest test_file::test_name` after it fails for debugging purposes, I have often found that what I end up doing is pulling that test out into a single file that I can run with none of the pytest machinery in the way. If it then depends on a bunch of fixtures this is painful because you have to find all those and so forth.

Reply:
I'll leave this conversation unresolved for now so that it's easy to find when we finally come back to this discussion.